### ISSUE TITLE
Veracrypt does not compile in embedded Debian-based system Linux 4.14.24-qnap x86_64

### ISSUE THREAD

**Issue by jmarks-lab** [veracrypt/VeraCrypt#895]
compiling the wxWidgets static library in Veracrypt on QNAP NAS
`# make WXSTATIC=1 WX_ROOT=/share/CACHEDEV1_DATA/jmarks/veracrypt_files/wxWidgets-3.1.5 ARCH=X86_64 CC=gcc WITHGTK3=1`

fails with the following:
```
Compiling cpu.c
In file included from ../Crypto/cpu.h:200:0,
                 from ../Crypto/cpu.c:3:
../Crypto/cpu.c: In function 'TryAESNI':
/share/CACHEDEV1_DATA/.qpkg/Entware/lib/gcc/x86_64-openwrt-linux-gnu/7.4.0/include/wmmintrin.h:61:1: error: inlining failed in call to always_inline '_mm_aesenc_si128': target specific option mismatch
 _mm_aesenc_si128 (__m128i __X, __m128i __Y)
 ^~~~~~~~~~~~~~~~
../Crypto/cpu.c:257:14: note: called from here
   ciphered = _mm_aesenc_si128(block, subkey);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../Crypto/cpu.h:200:0,
                 from ../Crypto/cpu.c:3:
/share/CACHEDEV1_DATA/.qpkg/Entware/lib/gcc/x86_64-openwrt-linux-gnu/7.4.0/include/wmmintrin.h:61:1: error: inlining failed in call to always_inline '_mm_aesenc_si128': target specific option mismatch
 _mm_aesenc_si128 (__m128i __X, __m128i __Y)
 ^~~~~~~~~~~~~~~~
../Crypto/cpu.c:257:14: note: called from here
   ciphered = _mm_aesenc_si128(block, subkey);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[1]: *** [/share/CACHEDEV1_DATA/jmarks/veracrypt_files/VeraCrypt_1.25.7_Source/src/Build/Include/Makefile.inc:21: ../Crypto/cpu.o] Error 1
```

This is the identical behavior reported years ago under Fedora 26 - issue #198. However, the solution then does not fix the problem.

Environment
hardware:  QNAP TVS-672XT NAS with X86_64 processor (Intel i5)
VeraCrypt version:1.25.7
wxWidgets 3.1.5

thank you

**Comment by idrassi** [comment]
To fix your issue, just set the variable `GCC_GTEQ_440 `to 1 in the Makefile instead of calculating it from gcc dumpversion.
So, it will be:

```
			GCC_GTEQ_440 := 1
			GCC_GTEQ_430 := 1
```
This will force the Makefile to fall to the part where `-maes` is added to CFLAGS and CXXFLAGS.

**Comment by jmarks-lab** [comment]
Thank you very much for getting back to me.
Editing the makefile as you suggested unfortunately gave me the same error.

Here is the edited section of the Makefile. 
I have likely misunderstood what to do.
	
```
ifeq "$(SIMD_SUPPORTED)" "1"
		CFLAGS += -msse2
		CXXFLAGS += -msse2


		GCC_GTEQ_440 := $(shell expr `$(CC) -dumpversion | sed -e 's/\.\([0-9][0-9]\)/\1/g' -e 's/\.\([0-9]\)/0\1/g' -e 's/^[0-9]\{3,4\}$$/&00/' -e 's/^[0-9]\{1,2\}$$/&0000/'` \>= 40400)
		GCC_GTEQ_430 := $(shell expr `$(CC) -dumpversion | sed -e 's/\.\([0-9][0-9]\)/\1/g' -e 's/\.\([0-9]\)/0\1/g' -e 's/^[0-9]\{3,4\}$$/&00/' -e 's/^[0-9]\{1,2\}$$/&0000/'` \>= 40300)

		GCC_GTEQ_440 := 1
		GCC_GTEQ_430 := 1

		ifeq "$(GCC_GTEQ_440)" "1"
			CFLAGS += -maes
			CXXFLAGS += -maes
		endif


		ifeq "$(GCC_GTEQ_430)" "1"
			ifeq "$(origin SSSE3)" "command line"
				CFLAGS += -mssse3
				CXXFLAGS += -mssse3
			endif

			ifeq "$(origin SSE41)" "command line"
				CFLAGS += -mssse3 -msse4.1
				CXXFLAGS += -mssse3 -msse4.1
			endif
		endif
	endif

```
Here is the compile error:
```
Compiling cpu.c
In file included from ../Crypto/cpu.h:200:0,
                 from ../Crypto/cpu.c:3:
../Crypto/cpu.c: In function 'TryAESNI':
/share/CACHEDEV1_DATA/.qpkg/Entware/lib/gcc/x86_64-openwrt-linux-gnu/7.4.0/include/wmmintrin.h:61:1: error: inlining failed in call to always_inline '_mm_aesenc_si128': target specific option mismatch
 _mm_aesenc_si128 (__m128i __X, __m128i __Y)
 ^~~~~~~~~~~~~~~~
../Crypto/cpu.c:257:14: note: called from here
   ciphered = _mm_aesenc_si128(block, subkey);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../Crypto/cpu.h:200:0,
                 from ../Crypto/cpu.c:3:
/share/CACHEDEV1_DATA/.qpkg/Entware/lib/gcc/x86_64-openwrt-linux-gnu/7.4.0/include/wmmintrin.h:61:1: error: inlining failed in call to always_inline '_mm_aesenc_si128': target specific option mismatch
 _mm_aesenc_si128 (__m128i __X, __m128i __Y)
 ^~~~~~~~~~~~~~~~
../Crypto/cpu.c:257:14: note: called from here
   ciphered = _mm_aesenc_si128(block, subkey);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[1]: *** [/share/CACHEDEV1_DATA/jmarks/veracrypt_files/VeraCrypt_1.25.7_Source/src/Build/Include/Makefile.inc:21: ../Crypto/cpu.o] Error 1

```

Thank you very much for your help.

**Comment by idrassi** [comment]
It should have worked if gcc enables AESNI but it seems that it doesn't enable it for your machine. 
What is your CPU? Does it have AESNI?

The simplest would be to disable AESNI using `CRYPTOPP_DISABLE_AESNI` define and for that I propose the following:

```
ifeq "$(SIMD_SUPPORTED)" "1"
		CFLAGS += -msse2 -DCRYPTOPP_DISABLE_AESNI
		CXXFLAGS += -msse2 -DCRYPTOPP_DISABLE_AESNI
endif
```

Does it work now?

**Comment by jmarks-lab** [comment]
Before I try this, I would like to make sure I am following your suggestion correctly. You would like me to replace this entire block:

```

ifeq "$(SIMD_SUPPORTED)" "1"
		CFLAGS += -msse2
		CXXFLAGS += -msse2
		GCC_GTEQ_440 := $(shell expr `$(CC) -dumpversion | sed -e 's/\.\([0-9][0-9]\)/\1/g' -e 's/\.\([0-9]\)/0\1/g' -e 's/^[0-9]\{3,4\}$$/&00/' -e 's/^[0-9]\{1,2\}$$/&0000/'` \>= 40400)
		GCC_GTEQ_430 := $(shell expr `$(CC) -dumpversion | sed -e 's/\.\([0-9][0-9]\)/\1/g' -e 's/\.\([0-9]\)/0\1/g' -e 's/^[0-9]\{3,4\}$$/&00/' -e 's/^[0-9]\{1,2\}$$/&0000/'` \>= 40300)

		ifeq "$(GCC_GTEQ_440)" "1"
			CFLAGS += -maes
			CXXFLAGS += -maes
		endif


		ifeq "$(GCC_GTEQ_430)" "1"
			ifeq "$(origin SSSE3)" "command line"
				CFLAGS += -mssse3
				CXXFLAGS += -mssse3
			endif

			ifeq "$(origin SSE41)" "command line"
				CFLAGS += -mssse3 -msse4.1
				CXXFLAGS += -mssse3 -msse4.1
			endif
                   endif
endif

```


with this?

```
feq "$(SIMD_SUPPORTED)" "1"

		CFLAGS += -msse2 -DCRYPTOPP_DISABLE_AESNI
		CXXFLAGS += -msse2 -DCRYPTOPP_DISABLE_AESNI

endif`
```

**Comment by idrassi** [comment]
Yes, that’s correct.

**Comment by jmarks-lab** [comment]
I am afraid that this did not work. The compile error looks identical:
`Compiling cpu.c
In file included from ../Crypto/cpu.h:200:0,
                 from ../Crypto/cpu.c:3:
../Crypto/cpu.c: In function 'TryAESNI':
/share/CACHEDEV1_DATA/.qpkg/Entware/lib/gcc/x86_64-openwrt-linux-gnu/7.4.0/include/wmmintrin.h:61:1: error: inlining failed in call to always_inline '_mm_aesenc_si128': target specific option mismatch
 _mm_aesenc_si128 (__m128i __X, __m128i __Y)
 ^~~~~~~~~~~~~~~~
../Crypto/cpu.c:257:14: note: called from here
   ciphered = _mm_aesenc_si128(block, subkey);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../Crypto/cpu.h:200:0,
                 from ../Crypto/cpu.c:3:
/share/CACHEDEV1_DATA/.qpkg/Entware/lib/gcc/x86_64-openwrt-linux-gnu/7.4.0/include/wmmintrin.h:61:1: error: inlining failed in call to always_inline '_mm_aesenc_si128': target specific option mismatch
 _mm_aesenc_si128 (__m128i __X, __m128i __Y)
 ^~~~~~~~~~~~~~~~
../Crypto/cpu.c:257:14: note: called from here
   ciphered = _mm_aesenc_si128(block, subkey);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[1]: *** [/share/CACHEDEV1_DATA/jmarks/veracrypt_files/VeraCrypt_1.25.7_Source/src/Build/Include/Makefile.inc:21: ../Crypto/cpu.o] Error 1
make: *** [Makefile:451: all] Error 2
`

The CPU in this NAS is a modern 4 core i5-8400T
Below is the output of 
`cat /proc/cpuinfo `
for the first core only. You will see that the aes flag is there

`processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 158
model name      : Intel(R) Core(TM) i5-8400T CPU @ 1.70GHz
stepping        : 10
microcode       : 0xd6
cpu MHz         : 3000.678
cache size      : 9216 KB
physical id     : 0
siblings        : 6
core id         : 0
cpu cores       : 6
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid dep_fpu_cs_ds mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass
bogomips        : 3408.00
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:
`
Many thanks for your help with this.

**Comment by idrassi** [comment]
Thank you for the update.
Clearly AES support should work but for some reason the compiler doesn't want to expose AES functions to us.

What I propose to you is to pull the latest source code from git and then use the following in the Makefile:

```
ifeq "$(SIMD_SUPPORTED)" "1"

		CFLAGS += -msse2 -maes
		CXXFLAGS += -msse2 -maes

endif
```

If it still doesn't work, then this means that OpenWrt compiler you are using doesn't enable AESNI or that the OpenWrt kernel doesn't have AESNI.

In this case, you should call make like this:

`make NOAESNI=1`

This is a new Makefile option that was added to master few days ago and it enables to completely remove AESNI support from VeraCrypt.

**Comment by idrassi** [comment]
The NOAESNI option of Makefile is included in latest release 1.25.9.
